sopel/modules/safety.py

```diff
@@ -110,129 +152,180 @@ def setup(bot):
                 # only publicly routable domains matter; skip loopback/link-local stuff
                 malware_domains.add(domain)
 
+    bot.memory["safety_cache_local"] = malware_domains
 
-def shutdown(bot):
+
+def shutdown(bot: Sopel):
     bot.memory.pop('safety_cache', None)
+    bot.memory.pop('safety_cache_local', None)
     bot.memory.pop('safety_cache_lock', None)
 
 
-def _download_domain_list(path):
-    url = 'https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts'
-    LOGGER.info('Downloading malicious domain list from %s', url)
-    urlretrieve(url, path)
-
-
 @plugin.rule(r'(?u).*(https?://\S+).*')
 @plugin.priority('high')
 @plugin.output_prefix(PLUGIN_OUTPUT_PREFIX)
-def url_handler(bot, trigger):
+def url_handler(bot: Sopel, trigger: Trigger):
     """Checks for malicious URLs"""
-    check = True    # Enable URL checking
-    strict = False  # Strict mode: kick on malicious URL
-    positives = 0   # Number of engines saying it's malicious
-    total = 0       # Number of total engines
-    use_vt = True   # Use VirusTotal
-    check = bot.config.safety.enabled_by_default
-    if check is None:
-        # If not set, assume default
-        check = True
-    # DB overrides config:
-    setting = bot.db.get_channel_value(trigger.sender, 'safety')
-    if setting is not None:
-        if setting == 'off':
-            return  # Not checking
-        elif setting in ['on', 'strict', 'local', 'local strict']:
-            check = True
-        if setting == 'strict' or setting == 'local strict':
-            strict = True
-        if setting == 'local' or setting == 'local strict':
-            use_vt = False
-
-    if not check:
-        return  # Not overridden by DB, configured default off
+    mode = bot.db.get_channel_value(
+        trigger.sender,
+        "safety",
+        bot.settings.safety.default_mode,
+    )
+    if mode == "off":
+        return
+    local_only = "local" in mode or bot.settings.safety.vt_api_key is None
+    strict = "strict" in mode
 
-    try:
-        netloc = urlparse(trigger.group(1)).netloc
-    except ValueError:
-        return  # Invalid IPv6 URL
+    for url in tools.web.search_urls(trigger):
+        safe_url = "hxx" + url[3:]
```

Comment:
This got me to check what `search_urls()` actually finds. The default list of schemes includes ftp:

https://github.com/sopel-irc/sopel/blob/2e25645487c87108330a992f68336a6b2da42188/sopel/tools/web.py#L251

Unless this plugin handles FTP links, it should search for URLs as below.

```suggestion
    for url in tools.web.search_urls(trigger, schemes=['http', 'https']):
        safe_url = "hxx" + url[3:]
```

Reply:
VT doesn't seem to take FTP links, but I don't mind running them through the hosts blocklist. Will .startswith() it